Fix playback sequence of episodes in a series/season

andrewrabert opened the pull request:
**Changes**
- Adds an option to a season's menu to `Play All From Here` (default behavior is to play only the specific season)
- Playback will start on the first unplayed episode, or at the beginning of the season
- Fixes the previous episode OSD button. (Ex. starting playback on episode 2 will now allow the previous button to navigate to episode 1).


**Issues**
Relates to https://github.com/jellyfin/jellyfin-web/pull/5415#pullrequestreview-2020486301

1hitsong commented:
> Adds an option to a season's menu to Play All From Here (default behavior is to play only the specific season)

Confirmed. I see the option directly under the Play option.

> Playback will start on the first unplayed episode, or at the beginning of the season

Setup:
EP1: Unwatched
EP2: Watched

Outcome:
Playback starts with EP1

---
Setup:
EP1: Watched
EP2: Unwatched

Outcome:
Playback starts with EP2

---
Setup:
EP1: Watched
EP2: Watched

Outcome:
Playback starts with EP1

> Fixes the previous episode OSD button. (Ex. starting playback on episode 2 will now allow the previous button to navigate to episode 1).

Setup:
EP1: Watched
EP2: Unwatched

If I use the Play from here option, playback begins with EP2 and I can click the back button in the OSD to play EP1.

If I go into the season and press play on EP1, I can click the forward button to play EP2. I can then click the back in the OSD to play EP1.

However, if I go into the season and press play on EP2, if I click the back in the OSD nothing happens.

Is this expected behavior keeping this PR in mind?

nvllsvm commented:
> If I go into the season and press play on EP1, I can click the forward button to play EP2. I can then click the back in the OSD to play EP1.
> 
> However, if I go into the season and press play on EP2, if I click the back in the OSD nothing happens.

Fixed with https://github.com/jellyfin/jellyfin-web/pull/5423/commits/1586db13069829c9193cb5640fbdf9b24b3aa807

1hitsong commented:
> > However, if I go into the season and press play on EP2, if I click the back in the OSD nothing happens.
> 
> Fixed with [1586db1](https://github.com/jellyfin/jellyfin-web/commit/1586db13069829c9193cb5640fbdf9b24b3aa807)

✅ Confirmed that given this scenario, pressing the back button from direct playing episode 2 now moves playback to episode 1.

1hitsong commented on src/components/playback/playbackmanager.js:
```diff
@@ -1868,38 +1868,43 @@ class PlaybackManager {
-                        });
+                        }
+                        if (episodesResult.StartIndex === undefined) {
+                            episodesResult.StartIndex = startSeasonId ? seasonStartIndex : 0;
```
Is this line testing the correct variable? Should it be testing `seasonStartIndex`?

nvllsvm replied:
It should work as I had it, but I guess it's possible that `seasonStartIndex` could never get set for some reason. Changed to https://github.com/jellyfin/jellyfin-web/pull/5423/commits/8fcc38a7bcb8b6fd0751c6e6e43abf7027601eac

1hitsong replied:
All my tests still work as expected and this was the only line of code that gave me pause.

andrewrabert committed:
Fix+enhance playback sequence of episodes in a series/season
- Adds an option to a season's menu to `Play All From Here` (default
  behavior is to play only the specific season)
- Playback will start on the first unplayed episode, or at the beginning
  of the season
- After starting playback, navigation to prior episodes is immediately possible
  using the previous episode navigation button
- Fix previous button not navigating to previous episode when starting
  on a non-first episode (ex. episode 1 when starting playback on episode 2)
